R/S3_emmeans.R

```diff
@@ -0,0 +1,43 @@
+### Support for emmeans package
+#
+# Note: the recover_data and emm_basis methods are registered dynamically
+#   (see zzz.R). So these functions are not exported
+#
+# Apologies from RVL: I use "=" rather than "<-" for top-level assignment
+
+recover_data.lm_robust = function(object, ...) {
+  data = getS3method("recover_data", "lm")(object, ...)
+  if (object$rank < object$k)  # rank-deficient. Need to pass dataset to emm_basis
+    attr(data, "pass.it.on") = TRUE
+  data
+}
+
+emm_basis.lm_robust = function(object, trms, xlev, grid, ...) {
+  # coef() works right for lm but coef.aov tosses out NAs
+  bhat = coef(object)
+  n.mult = ifelse(is.matrix(bhat), ncol(bhat), 1)  # columns in mult response
+  m = suppressWarnings(model.frame(trms, grid, na.action = na.pass, xlev = xlev))
+  X = model.matrix(trms, m, contrasts.arg = object$contrasts)
+  V = emmeans::.my.vcov(object, ...)
+
+  if (sum(is.na(bhat)) > 0) {
+    desmat = model.matrix(trms, data = attr(object, "data"))
+    nbasis = estimability::nonest.basis(desmat)
+  }
+  else
+    nbasis = estimability::all.estble
```

Comment:
Place the smaller branch first. eg

    if (!anyNA(bhat)) {
        nbasis = estimability::all.estble
    } else {
        # handle nas
    }